### ISSUE TITLE
Single Target Scraping mode returns inconsistent results under load

### ISSUE THREAD

**Issue by vervas** [promregator/promregator#51]
## Summary
Using promregator on Single Target Scraping mode returns inconsistent series for requested targets under high load.

## Observed Behavior
We have ~30 services each with 3 instances so around 90 targets scraped on a 1m interval.  Prometheus then returns quite frequently the message "Error on ingesting out-of-order samples" and drops the scraped metrics. Looking deeper I have observed that under load, promregator returns series corresponding to different targets than the one queried.

For example if I hit `https://promregator:443/singleTargetMetrics/XXX/0` I get series like `promregator_up{cf_instance_id="YYY:1",cf_instance_number="1",...} 1.0`, while others are just working fine.

This results in both duplicate series and missing ones thus having inconsistent metrics.

## Expected Behavior
Hitting `https://promregator:443/singleTargetMetrics/XXX/0` I would expect to get series for `cf_instance_id` "XXX" like `promregator_up{cf_instance_id="XXX:0",cf_instance_number="0",...} 1.0`.

## Environment
The environment is the following:
- promregator version: 0.3.0
- runtime: docker image running on PCF
- cf api version: 2.94.0
- number of defined targets: 29
- number of targets from /discovery: 88
- scraping interval: 1m

The issue occurs with either one instance of promregator running or many, though scaling out reduces the amount of errors.

**Comment by eaglerainbow** [comment]
Thanks for reporting the issue - sounds like some missing lock or mixing up of instances, which is ugly.

Do I get you right that you cannot reproduce the issue using Single Endpoint Scraping mode? (it's not that I want you to switch to that mode, but it might help to track down the source of the error)

**Comment by eaglerainbow** [comment]
One more question: Can you please check if you experience the mix up only at metrics like `promregator_up` or also at metrics, which are coming from the scraped target? That would also help to reduce the scope for the error analysis.

Thanks!

**Comment by eaglerainbow** [comment]
Looking at the code, I currently suspect https://github.com/promregator/promregator/blob/d5ad8cc0c570412c161f43785e83b6112797a576/src/main/java/org/cloudfoundry/promregator/fetcher/MetricsFetcherMetrics.java#L63
(but I don't have a clue, yet, why this could go wrong). 

On your requests for `https://promregator:443/singleTargetMetrics/XXX/0`, are you able to see only
```
promregator_up{cf_instance_id="YYY:1",cf_instance_number="1",...} 1.0
``
but also both and
```
promregator_up{cf_instance_id="YYY:0",cf_instance_number="0",...} 1.0
promregator_up{cf_instance_id="YYY:0",cf_instance_number="0",...} 1.0
```
That could give yet another piece of the puzzle.

**Comment by eaglerainbow** [comment]
Bingo! I found it.
https://github.com/promregator/promregator/blob/d5ad8cc0c570412c161f43785e83b6112797a576/src/main/java/org/cloudfoundry/promregator/endpoint/AbstractMetricsEndpoint.java#L44 is being ignored by Spring Framework and for each request the same instance is called by different threads. That is not as it has to be.

**Comment by eaglerainbow** [comment]
@vervas: I just released [version 0.3.1](https://github.com/promregator/promregator/releases/tag/v0.3.1), which I hope that it contains a fix for this issue here. It'll be great if you could confirm that, as I am lacking an environment to properly test this right now.

Even if this solved your issue, please keep this issue here open. I'd like to use it as a reminder for thinking about how we can improve our test infrastructure to ensure that such kind of issues don't re-appear (currently, the test infrastructure isn't capable of detecting this - most likely it goes into the direction of #37).

**Comment by vervas** [comment]
Hey, thanks for jumping straight on that and releasing a new version already.

As you said this looks like a concurrency issue. I assume you figured it already by you your previous message. Nonetheless to answer to your previous questions:
- The issue occurs only on Single Target Scraping mode
- It occurs for all the metrics and not just `promregator_up`
- When I get an inconsistent result (which is not the case for every request) the URL might point to say `/XXX/0` and the response would provide *consistent* results for all the metrics of a different target say `YYY:0` but only the series with that particular `cf_instance_id`. So no mixed results in terms of tagging on a single target.
- At the same time there are no duplicate series on a single response, but since different targets might return the same series, I assume this is where the conflict occurs.

I will try it actually straight away and get back with feedback.

**Comment by eaglerainbow** [comment]
Thanks! As far as I can understand your observations, it would fit to what I could imagine may happen, if https://github.com/promregator/promregator/blob/d5ad8cc0c570412c161f43785e83b6112797a576/src/main/java/org/cloudfoundry/promregator/endpoint/AbstractMetricsEndpoint.java#L44 did not work correctly. So, in short, I think we have a good chance, that my correction will do the trick...

**Comment by eaglerainbow** [comment]
mmh.. I just browsed through it: could be, but isn't the case automatically. 
Are you able to paste here a copy of the corresponding target definition in *Prometheus* ? Do you have any recording rules active?

**Comment by vervas** [comment]
No recording rules. By target definition you are referring to the response from the `/discovery` endpoint or something else?

The error `Error on ingesting out-of-order samples` occurs for every single target on every scrape accompanied with `num_dropped=1`

**Comment by eaglerainbow** [comment]
I think I see your problem. As this quite distinct to the current issue discussed here, I have opened #52 for your case. Let's continue there on this.

**Comment by eaglerainbow** [comment]
Internal remark: unit test available on `master` (as of `0.4.0-SNAPSHOT`) for this case. Not downported to `rel-0.3`.
